### PR TITLE
SONARIAC-966 ARM Bicep support: create unaryOperator and multilineString

### PR DESCRIPTION
**Pull request by mstachniuk**

### CODE REVIEW

File: iac-common/src/main/java/org/sonar/iac/common/parser/grammar/Punctuator.java [hunk 1]
```diff
@@ -56,7 +56,8 @@ public enum Punctuator implements GrammarRuleKey {
   PERCENT_LCURLY_TILDE("%{~"),
   TILDE_RCURLY("~}"),
   DOLLAR("$"),
-  APOSTROPHE("'");
+  APOSTROPHE("'"),
```
Comment (jonas-wielage-sonarsource) [inline iac-common/src/main/java/org/sonar/iac/common/parser/grammar/Punctuator.java:59]: I don't see any use of a single `APOSTROPHE` if I'm not mistaken. Can be deleted IMO.

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepLexicalGrammar.java [hunk 2]
```diff
@@ -95,7 +98,13 @@ public enum BicepLexicalGrammar implements GrammarRuleKey {
...
   IMPORT_AS_CLAUSE,
   IMPORT_WITH_CLAUSE,
-  AMBIENT_TYPE_REFERENCE_VALUE;
+  AMBIENT_TYPE_REFERENCE_VALUE,
+  UNARY_OPERATOR_VALUE;
+
+  private static final List<Punctuator> LIST_OF_PUNCTUATORS = List.of(
+    Punctuator.EQU, Punctuator.COLON, Punctuator.LCURLYBRACE, Punctuator.RCURLYBRACE, Punctuator.APOSTROPHE, Punctuator.COMMA,
+    Punctuator.RPARENTHESIS, Punctuator.LPARENTHESIS, Punctuator.LBRACKET, Punctuator.RBRACKET,
+    Punctuator.DOLLAR_LCURLY, Punctuator.TRIPLE_APOSTROPHE);
```
Comment (jonas-wielage-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepLexicalGrammar.java:107]: Good addition!

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepLexicalConstant.java [hunk 3]
```diff
@@ -43,6 +43,8 @@ public class BicepLexicalConstant {
   public static final String QUOTED_STRING_LITERAL_NO_QUOTES = "(?:(?!'|\\$\\{)[\\s\\S])*+";
   public static final String IDENTIFIER_LITERAL = "[a-zA-Z_][a-zA-Z_0-9]*+";
   public static final String AMBIENT_TYPE = "(?:array|bool|int|object|string)(?![\\S])";
+  public static final String UNARY_OPERATOR = "!|-|\\+";
+  public static final String MULTILINE_STRING = "(?:(?!''')(?:[\\s\\S])?+)*+";
```
Comment (jonas-wielage-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepLexicalConstant.java:47]: I'm always fascinated about this regex magic. Nice to exclude previous triple apostrophe.

Comment (jonas-wielage-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepLexicalConstant.java:47]: I feel like having `BicepLexicalGrammar.MULTILINE_STRING` and `BicepLexicalConstant.MULTILINE_STRING` share the same name is a bit confusing. (Because a `BicepLexicalConstant.MULTILINE_STRING` is part of a `BicepLexicalGrammar.MULTILINE_STRING_REGEX` which is part of a `BicepLexicalGrammar.MULTILINE_STRING`)


 I would suggest to renaming something.
Maybe  `BicepLexicalConstant.MULTILINE_STRING` to  `BicepLexicalConstant.MULTILINE_STRING_CONTENT`

File: iac-extensions/arm/src/test/java/org/sonar/iac/arm/tree/impl/bicep/MultilineStringImplTest.java [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+      "a",
+      "123",
+      "BBB",
+      "'''");
+    MultilineString tree = parse(code, BicepLexicalGrammar.MULTILINE_STRING);
+    Assertions.assertThat(tree.value()).isEqualTo("a\n123\nBBB\n");
+    assertThat(tree.getKind()).isEqualTo(ArmTree.Kind.MULTILINE_STRING);
+    Assertions.assertThat(tree.children())
+      .map(c -> ((SyntaxToken) c).value())
+      .containsExactly("'''", "a\n123\nBBB\n", "'''");
```
Comment (jonas-wielage-sonarsource) [inline iac-extensions/arm/src/test/java/org/sonar/iac/arm/tree/impl/bicep/MultilineStringImplTest.java:102]: Feel free to use ArmTestUtils :)

File: iac-common/src/test/java/org/sonar/iac/common/testing/IacTestUtils.java [hunk 5]
```diff
@@ -41,7 +41,7 @@ private IacTestUtils() {
   }
 
   public static String code(String... lines) {
-    return StringUtils.join(lines, System.getProperty("line.separator"));
+    return StringUtils.join(lines, "\n");
```
Comment (jonas-wielage-sonarsource) [inline iac-common/src/test/java/org/sonar/iac/common/testing/IacTestUtils.java:44]: Why this change?

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepLexicalGrammar.java [hunk 6]
```diff
@@ -95,7 +98,13 @@ public enum BicepLexicalGrammar implements GrammarRuleKey {
 
   IMPORT_AS_CLAUSE,
   IMPORT_WITH_CLAUSE,
-  AMBIENT_TYPE_REFERENCE_VALUE;
+  AMBIENT_TYPE_REFERENCE_VALUE,
+  UNARY_OPERATOR_VALUE;
+
+  private static final List<Punctuator> LIST_OF_PUNCTUATORS = List.of(
+    Punctuator.EQU, Punctuator.COLON, Punctuator.LCURLYBRACE, Punctuator.RCURLYBRACE, Punctuator.APOSTROPHE, Punctuator.COMMA,
```
Comment (jonas-wielage-sonarsource) [inline iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepLexicalGrammar.java:105]: `Punctuator.APOSTROPHE` is not needed I think

**Review by jonas-wielage-sonarsource — APPROVED** [review]
LGTM! Please see my comment before merging.
Also I think you have to implement a basic PunctuatorsTest to satisfy sonar analysis.

File: iac-common/src/main/java/org/sonar/iac/common/parser/grammar/Punctuator.java (hunk 1 above)
Comment (mstachniuk) [reply iac-common/src/main/java/org/sonar/iac/common/parser/grammar/Punctuator.java:59]: It's used in `STRING_COMPLETE`, `INTERPOLATED_STRING_LEFT_PIECE`, and other places, but not in this PR.
Here the change is from semicolon to colon to add new Punctuators.

File: iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepLexicalConstant.java (hunk 3 above)
Comment (mstachniuk) [reply iac-extensions/arm/src/main/java/org/sonar/iac/arm/parser/bicep/BicepLexicalConstant.java:47]: It was tricky

File: iac-common/src/test/java/org/sonar/iac/common/testing/IacTestUtils.java (hunk 5 above)
Comment (mstachniuk) [reply iac-common/src/test/java/org/sonar/iac/common/testing/IacTestUtils.java:44]: Because it was not compatible with `TreeModelTest` (a basis of `BicepTreeModelTest`).
It failed, but only on Windows machines :-( 
See: https://github.com/SonarSource/sonar-iac/blob/947e23be421244f38c6867d8d19e044438eebccf/iac-common/src/test/java/org/sonar/iac/common/testing/TreeModelTest.java#L117

File: iac-common/src/main/java/org/sonar/iac/common/parser/grammar/Punctuator.java (hunk 1 above)
Comment (jonas-wielage-sonarsource) [reply iac-common/src/main/java/org/sonar/iac/common/parser/grammar/Punctuator.java:59]: Ah I missed that during the review.
